src/main/scala/tsiHost/TsiHostWidgetUnitTest.scala

```diff
@@ -0,0 +1,172 @@
+  val hostTSIHostWidget = LazyModule(new TLTSIHostWidget(beatBytes))
+
+  // lives in target-land (connects to the TSIHost widget in host-land)
+  val targetSerdes = LazyModule(new TLSerdesser(
+    w = serWidth,
+    clientParams = TLClientParameters(
+      name = "tl-desser",
+      sourceId = IdRange(0, numXacts)),
+    managerParams = TLManagerParameters(
+      address = Seq(addrSet),
+      regionType = regionTyp,
+      supportsGet = TransferSizes(1, lineBytes),
+      supportsPutFull = TransferSizes(1, lineBytes),
+      supportsPutPartial = TransferSizes(1, lineBytes)),
+    beatBytes = beatBytes))
+
+  // rom living in target-land. connects to the client of the serdes
+  val targetRomData = Seq(
+    BigInt("7766554433221100", 16),
+    BigInt("FFEEDDCCBBAA9988", 16),
+    BigInt("0123456789ABCDEF", 16),
+    BigInt("FEDCBA9876543210", 16),
+    BigInt("76543210FDECBA98", 16))
+  val targetRom = LazyModule(new TLROM(0, 64,
+    targetRomData.flatMap(
+      data => (0 until 8).map(i => ((data >> (i * 8)) & 0xff).toByte)),
+    beatBytes = beatBytes))
+
+  // TLFragmenter splits up the multi beat packet into single beats that the left hand side can take
+  // (of size beatBytes).
+  targetRom.node := TLFragmenter(beatBytes, lineBytes) := TLBuffer() := targetSerdes.clientNode
+
+  // fuzzer that lives in target-land that sends random requests to the ram on the host
+  val targetFuzzer = LazyModule(new TLFuzzer(
+    nOperations = 32,
+    inFlight = numXacts))
+
+  // ram living in host-land. for the fuzzer to read and write from
+  val hostRam = LazyModule(new TLTestRAM(
+    address = addrSet,
+    beatBytes = beatBytes))
+
+  // connect fuzzer to the target serdes
+  targetSerdes.managerNode := TLBuffer() := targetFuzzer.node
+  // connect the host node to the host rom
+  hostRam.node := TLFragmenter(beatBytes, lineBytes) := TLBuffer() := hostTSIHostWidget.externalClientNode
+
+  // implementation of the module
+  lazy val module = new LazyModuleImp(this) {
+    // i/o to connect to the unit test interface
+    val io = IO(new Bundle with UnitTestIO)
+
+    // TSI cmd of the form (cmd, lowAddr, highAddr, lowLen, highLen)
+    // note that len is actually the len - 1 (e.g. len of 4 is really 5)
+    // and len is in increments of 32b
+    // O is read, 1 is write
+    val inputTsiSeq = Seq(Seq(0,  0x0, 0, 0x0, 0),
+                          Seq(0,  0x4, 0, 0x0, 0),
+                          Seq(0,  0xC, 0, 0x0, 0),
+                          Seq(0, 0x10, 0, 0x0, 0),
+                          Seq(0, 0x20, 0, 0x1, 0))
+
+    // 32b sequences expected from the TSI response
+    val outputTsiData = Seq(Seq(BigInt("33221100", 16)),
+                            Seq(BigInt("77665544", 16)),
+                            Seq(BigInt("FFEEDDCC", 16)),
+                            Seq(BigInt("89ABCDEF", 16)),
+                            Seq(BigInt("FDECBA98", 16), BigInt("76543210", 16)))
```

Comment:
This seems fragile, this should be some function that derives the responses from the inputTsiSeq and the known state, rather than a hardcoded thing. Having the TSICommand objects like mentioned above might make this easier.